Up front extrapolation/substitution

tokee opened the pull request:
Extrapolation/substitution in `kb-util`'s `YAML` expands system properties, arguments and YAML-paths. The previous implementation performed substitution at request time, which made it difficult to implement things like the [Visitor Pattern](https://en.wikipedia.org/wiki/Visitor_pattern) extraction of sub-structures without introducing errors.

The primary implementation in this pull request is to perform a full extrapolation/substitution across the whole YAML when `setExtrapolate(true)` is called. 

Secondary fixes are
* Cleanup of Javadoc by moving `yPath` description to the class javadoc
* Changed behaviour of `getMultiple(String)` to treat the input af an `yPath` to mirror `get`. This is a breaking change, which has been discussed with @jorntx. The break was deemed acceptable as the old `getMultiple(String)` has only been part of `YAML` for a month
* `List<YAML> sub = myyaml.getList<YAML>(yPath)` now works. Previously it was necessary to call an explicit method to get a list of YAMLS: `List<YAML> sub = myyaml.getYAMLList(yPath)`
* If `get(yPath)` finds more than 1 value, an exception is thrown. The logic is that `get` is singular and multiple matches are nearly always a sign of something gone wrong. If the caller really want to use the first value from an `yPath` and is fine with ignoring subsequent values, `getMultiple(yPath)` can be used
* Extrapolation now supports indirect resolving, e.g. calling `get("foo")` for the structure below will return the user name
```yaml
foo: ${path:bar}
bar:${path:boom}
boom: ${user.name}
```

Discussion point
* The extrapolation is not done automatically. This is the same behaviour as before, but might not be the best. Should it be enabled per default?

VictorHarbo commented on src/main/java/dk/kb/util/yaml/YAML.java:
```diff
@@ -146,16 +162,33 @@ public YAML(Map<String, Object> map, boolean extrapolateSystemProperties) {
+     * <p>
+     * Note: Dots {@code .} in YAML keys can be escaped with quotes: {@code foo.'a.b.c' -> [foo, a.b.c]}.
+     * <p>
+     * This method is equal to {@link #getSubMap(String)}. {@code getYAML} is preferred dut to clearer semantics.
```
Small typo -> due, not dut

VictorHarbo commented on src/main/java/dk/kb/util/yaml/YAML.java:
```diff
@@ -570,27 +617,30 @@ public String getString(String path, String defaultValue) throws NullPointerExce
-        visit(correctedPath, this, visitor);
-        return visitor.extractedValues;
-    }
+    // TODO: Evaluate the redefinition of getMultiple
+    // The old getMultiple 
+
+//    /**
+//     * Resolves all values related to a given key, from a YAML structure. All values that have the specified key are
+//     * returned as part of a list.
+//     * @deprecated
+//     * This is no longer the best method to visit values. Use {@link #visit(Object, YAML, YAMLVisitor)} instead.
+//     * @param key the key to look for in the input YAML.
+//     * @return a list of all values that can be cast to strings.
+//     */
+//    @Deprecated
+//    public List<Object> getMultiple(String key){
+//        String correctedPath = "**." + key;
+//        MultipleValuesVisitor visitor = new MultipleValuesVisitor();
+//        visit(correctedPath, this, visitor);
+//        return visitor.extractedValues;
+//    }
```
Please remove the old implementation completely. I agree with you and @jorntx on the change.

VictorHarbo commented on src/main/java/dk/kb/util/yaml/YAML.java:
```diff
@@ -1433,14 +1546,25 @@ public enum MERGE_ACTION {
+            for (StringSubstitutor substitutor : getSubstitutors()) {
+                current = current == null ? null : substitutor.replace(current);
+            }
+            if (++iterations == 10) {
```
New syntax for me here! Smart one!

VictorHarbo reviewed (changes requested):
Nicely done! 
I am trying to think of situations where you don't want the extrapolation to be executed and I have a hard time thinking of situations where this is the case. Therefore I believe it would make sense if we switched the default to true.

tokee commented:
> Nicely done! I am trying to think of situations where you don't want the extrapolation to be executed and I have a hard time thinking of situations where this is the case. Therefore I believe it would make sense if we switched the default to true.

I have switched this, but to maintain the option of **not** extrapolating, a lot of constructors & methods had to be overloaded with an `extrapolateSystemProperties` boolean argument.

VictorHarbo commented on src/main/java/dk/kb/util/yaml/YAML.java:
```diff
@@ -123,31 +119,56 @@ public YAML() {
      * @param map a map presumable delivered by SnakeYAML.
-     * @param extrapolateSystemProperties should system properties be extrapolated in values
+     * @param extrapolateSystemProperties whether system properties should be extrapolated in values.
+     *                                    Note that extrapolation cannot be turned off once enabled.
```
I really like these notes in the documentation!

VictorHarbo reviewed (approved):
This looks very good now. Merging into the branch as agreed upon last week.
